Feature log file wildcard support

bernd opened the pull request:
This includes a fair amount of refactoring of the internals. Especially the file observer.

Fixes #24

bernd committed:
Start a file tracking list to resolve paths with globs.

bernd committed:
Start a test for wildcard/globbing edge cases.

bernd committed:
Convert paths to absolute path in FileTrackingList.

bernd committed:
Remove warning about non-existent files, that can be normal with globbing.

bernd committed:
Add getRootPath() to FileTrackingList.

bernd committed:
Rename FileTrackingList to PathSet and move to file package.

bernd committed:
Add isInSet() to PathSet.

bernd committed:
Add equals() and hashCode() to PathSet.

bernd committed:
Add getter for the PathSet pattern.

bernd committed:
First working version with wildcard support.

This introduces a lot of internal changes:

- Introduce a PathSet that returns all files which match the pattern.
- Change FileObserver API to watch the root directory of a PathSet.
- Remove name strategy matching for now because it was hardcoded and got
  in the way. It will be restored once we make the name strategies
  configurable.
- Use one FsChangeListener per PathSet instead of per file.
- Change FileReaderService to take a PathSet instead of a Set<Path>.

bernd committed:
Add some javadoc to PathSet.

bernd committed:
Split path manually to make it work on Windows.

bernd committed:
Escape backslashes in the pattern on Windows to unbreak matching.

joschi commented on src/test/java/org/graylog/collector/file/PathSetTest.java:
```diff
@@ -0,0 +1,164 @@
+        final PathSet list = new PathSet("/var/log/syslog", new PathSet.FileTreeWalker() {
+            @Override
+            public void walk(Path basePath, FileVisitor<Path> visitor) throws IOException {
+                System.out.println("YO");
```
Yo!

bernd replied:
Yo!

joschi commented on src/test/java/org/graylog/collector/file/FileObserverTest.java:
```diff
@@ -288,8 +291,11 @@ public void cannotObservePath(Path path) {
         };
 
         // Make sure to use a non-existent path for the naming strategy here!
+        /*
```
These lines should be removed instead of being commented.

joschi commented on src/main/java/org/graylog/collector/file/FileObserver.java:
```diff
@@ -103,20 +91,19 @@ public FileObserver(WatchService watchService) {
-        keys.get(key).addEntry(new WatchEntry(path, namingStrategy, listener));
+
+        keys.putIfAbsent(key, Sets.<WatchPath>newConcurrentHashSet());
+        keys.get(key).add(new WatchPath(pathSet, listener));
```
Should initializing the key with an empty `ConcurrentHashSet` and adding the actual `WatchPath` be atomic?

If there are multiple threads accessing this object/method, we should guard this initialization and adding the `WatchPath` with a lock (or `synchronized(keys)`) because the thread might be suspended after `keys.get(key)`.

bernd replied:
Good, point. Will add a `synchronized` block.

joschi commented on src/main/java/org/graylog/collector/file/FileObserver.java:
```diff
@@ -142,32 +128,28 @@ protected void run() throws Exception {
-                    if (ev.kind() == ENTRY_MODIFY) {
-                        watchEntry.listener.pathModified(path);
+                        if (log.isTraceEnabled()) {
+                            log.trace("Ignoring change {} to path {} - No match in path set {}", ev.kind().name(), path.toString(), watchPath);
```
Calling `toString()`is unnecessary with the parameterized logging.

joschi commented on src/main/java/org/graylog/collector/file/FileReaderService.java:
```diff
@@ -126,53 +137,40 @@ protected void doStop() {
-
-        public FsChangeListener(Path monitoredFile) {
-            this.monitoredFile = monitoredFile;
+        public FsChangeListener() {
```
Unneeded default constructor.

bernd committed:
Guard operations on keys map with a synchronized block.

joschi commented on src/main/java/org/graylog/collector/file/PathSet.java:
```diff
@@ -0,0 +1,195 @@
+    public PathSet(final String pattern, FileTreeWalker fileTreeWalker) {
+        this.pattern = pattern;
+        this.fileTreeWalker = fileTreeWalker;
+        final String patternString = Utils.isWindows() ? pattern.replace("\\", "\\\\") : pattern;
```
Wouldn't this replace "C:\\Foobar" with "C:\\\\Foobar"? Does it matter if this happens or is it ignored?

joschi replied:
Maybe use [`Pattern#quote(String)`](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html#quote%28java.lang.String%29) instead?

bernd replied:
I have to test that on Windows and Linux. Not sure if this works for the globbing type:

`Pattern.quote("/foo/bar.log") => "\Q/foo/bar.log\E"`

bernd replied:
No, does not work. Using `Pattern.quote("C:\test\*.log")` creates a regex like this: `^QC:test\*\.logE$` That's not going to work.

The globbing code does something special...

joschi commented on src/main/java/org/graylog/collector/file/PathSet.java:
```diff
@@ -0,0 +1,195 @@
+
+    @Override
+    public int hashCode() {
+        return pattern.hashCode();
```
This will fail with an NPE if `pattern` is `null` (which isn't checked in the constructor).

joschi commented on src/main/java/org/graylog/collector/file/PathSet.java:
```diff
@@ -0,0 +1,195 @@
+
+        PathSet pathSet = (PathSet) o;
+
+        return pattern.equals(pathSet.pattern);
```
This will fail with an NPE if `pattern` is `null` (which isn't checked in the constructor).

bernd committed:
Smaller cleanups after PR comments.

- Remove commented code.
- No need for toString() in logging parameter.
- Remove empty default constructor.
- Remove stray println() in test.

bernd committed:
Add checkNotNull() for pattern in PathSet constructor.

bernd committed:
Simplify checkNotNull() call in PathSet.

joschi commented:
LGTM. :+1:
